src/main/java/org/opensearch/knn/jni/JNIService.java

```diff
@@ -13,28 +13,110 @@
 
 import org.apache.commons.lang.ArrayUtils;
 import org.opensearch.common.Nullable;
-import org.opensearch.knn.index.util.IndexUtil;
-import org.opensearch.knn.index.query.KNNQueryResult;
+import org.opensearch.knn.common.KNNConstants;
 import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.query.KNNQueryResult;
+import org.opensearch.knn.index.util.IndexUtil;
 
+import java.util.Locale;
 import java.util.Map;
 
 /**
  * Service to distribute requests to the proper engine jni service
  */
 public class JNIService {
+    /**
+     * Initialize an index for the native library. Takes in numDocs to
+     * allocate the correct amount of memory.
+     *
+     * @param numDocs    number of documents to be added
+     * @param dim        dimension of the vector to be indexed
+     * @param parameters parameters to build index
+     * @param knnEngine  knn engine
+     * @return address of the index in memory
+     */
+    public static long initIndex(long numDocs, int dim, Map<String, Object> parameters, KNNEngine knnEngine) {
+        if (KNNEngine.FAISS == knnEngine) {
+            if (IndexUtil.isBinaryIndex(knnEngine, parameters)) {
+                return FaissService.initBinaryIndex(numDocs, dim, parameters);
+            } else {
+                return FaissService.initIndex(numDocs, dim, parameters);
+            }
+        }
+
+        throw new IllegalArgumentException(
+            String.format(Locale.ROOT, "initIndexFromScratch not supported for provided engine : %s", knnEngine.getName())
+        );
+    }
+
+    /**
+     * Inserts to a faiss index.
+     *
+     * @param docs           ids of documents
+     * @param vectorsAddress address of native memory where vectors are stored
+     * @param dimension      dimension of the vector to be indexed
+     * @param parameters     parameters to build index
+     * @param indexAddress   address of native memory where index is stored
+     * @param knnEngine      knn engine
+     */
+    public static void insertToIndex(
+        int[] docs,
+        long vectorsAddress,
+        int dimension,
+        Map<String, Object> parameters,
+        long indexAddress,
+        KNNEngine knnEngine
+    ) {
+        int threadCount = (int) parameters.getOrDefault(KNNConstants.INDEX_THREAD_QTY, 0);
```

Comment:
why default is 0 here?

Reply:
https://github.com/opensearch-project/k-NN/pull/1840/files#diff-892c5ae9158d35b9f5fa9b67d53d7373ca6d2bf04d1052add99c329d03645409R70